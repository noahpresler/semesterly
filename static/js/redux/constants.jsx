import { getSchool, getSemester } from './init.jsx';
export const VALID_SCHOOLS = ["uoft", "jhu", "umd", "uo", "rutgers", "queens"];
export const VALID_SEMESTERS = ["F", "S"];
export const SET_SCHOOL = "SET_SCHOOL";
export const SET_SEMESTER = "SET_SEMESTER";
export const REQUEST_TIMETABLES = "REQUEST_TIMETABLES";
export const RECEIVE_TIMETABLES = "RECEIVE_TIMETABLES";
export const HALF_HOUR_HEIGHT = 25;
export const MAX_TIMETABLE_NAME_LENGTH = 30; // Length of "This is my fucking timetable!!"

/* server endpoints */
export const getCourseInfoEndpoint = (course_id) => {
	return "/courses/" + getSchool() + "/" + getSemester() + "/id/" + course_id + "/";
};
export const getCourseSearchEndpoint = (query) => {
	return "/search/" + getSchool() + "/" + getSemester() + "/" + query + "/";
};
export const getAdvancedSearchEndpoint = (query) => {
	return "/advanced_search/";
};
export const getTimetablesEndpoint = () => {
	return "/get_timetables/";
};
export const getUserInfoEndpoint = () => {
	return "/user/info/";
};
export const getSaveTimetableEndpoint = () => {
	return "/user/save_timetable/";
};
export const getSaveSettingsEndpoint = () => {
	return "/user/save_settings/"
}
export const getClassmatesEndpoint = () => {
	return "/user/get_classmates/"
}
export const getSchoolInfoEndpoint = () => {
	return "/school_info/" + getSchool() + "/";
}
export const getReactToCourseEndpoint = () => {
  return "/react/";
}
export const getLoadSavedTimetablesEndpoint = (semester) => {
  return "/user/get_saved_timetables/" + getSchool() + "/" + semester + "/";
}
export const getRequestShareTimetableLinkEndpoint = () => {
  return "/share/link/";
}

export const getSchoolSpecificInfo = (school) => {
	switch(school) {
		case "uoft":
			return {
        primaryDisplay: "code",
        areasName: "Breadths",
        departmentsName: "Departments",
        levelsName: "Levels",
        timesName: "Times",
        semesters: {
          F: "Fall 2016",
          S: "Winter 2017"
        },
        campuses: {
          1: "UTSG",
          2: "UTSC",
          5: "UTM"
        }
      }
		case "jhu":
			return {
        primaryDisplay: "name",
        areasName: "Areas",
        departmentsName: "Departments",
        levelsName: "Levels",
        timesName: "Times",
        semesters: {
          F: "Fall 2016",
          S: "Spring 2017"
        },
        campuses: {
          1: ""
        }
      }
		case "umd":
			return {
        primaryDisplay: "name",
        areasName: "Areas",
        departmentsName: "Departments",
        levelsName: "Levels",
        timesName: "Times",
        semesters: {
          F: "Fall 2016",
          S: "Spring 2017"
        },
        campuses: {
          1: ""
        }
      }
		default:
			return {
        primaryDisplay: "code",
        areasName: "Areas",
        departmentsName: "Departments",
        levelsName: "Levels",
        timesName: "Times",
        semesters: {
          F: "Fall 2016",
          S: "Spring 2017"
        },
        campuses: {
          1: ""
        }
      }
	}
}

export const COLOUR_DATA = [
    // red
    { background: "#FD7473", highlight: "#e36867", border: "#b15150", font: "#222" },
    // blue
    { background: "#5CCCF2", highlight: "#52b7d9", border: "#408ea9", font: "#222" },
    // turquoise 
    { background: "#36DEBB", highlight: "#30c7a8", border: "#259b82", font: "#222" },
    // 'yellow'
    { background: "#FFD462", highlight: "#e5be58", border: "#b29444", font: "#222" },
    // purple
    { background: "#C585DE", highlight: "#b177c7", border: "#895d9b", font: "#222" },
    // green
    { background: "#53e997", highlight: "#4ad187", border: "#3aa369", font: "#222" },
    { background: "#D4DBC8", highlight: "#B5BFA3", border: "#97A086", font: "#222" },
    { background: "#E7F76D", highlight: "#C9E20A", border: "#AFC11F", font: "#222" },
    { background: "#A3F5F2", highlight: "#7CD2CF", border: "#53ABA8", font: "#222" },
    { background: "#7499A2", highlight: "#668B94", border: "#6C7A89", font: "#222" },
    { background: "#E7F76D", highlight: "#C4D44D", border: "#6C7A89", font: "#222" },
    { background: "#C8F7C5", highlight: "#C4D44D", border: "#548A50", font: "#222" }
] // consider #CF000F, #e8fac3, #C8F7C5

export const REACTION_MAP = {
  FIRE: {
  	unicode: '\uD83D\uDD25',
  	name: 'Must Take'
  },
  LOVE: {
  	unicode: '\uD83D\uDE0D',
  	name: 'Love It'
  },
  CRAP: {
  	unicode: '\uD83D\uDCA9',
  	name: 'Crap Class'
  },
  OKAY: {
  	unicode: '\uD83D\uDE10',
  	name: "It's Aight"
  },
  BORING: {
  	unicode: '\uD83D\uDE34',
  	name: 'Boring'
  },
  HARD: {
  	unicode: '\uD83D\uDE29',
  	name: 'Hard'
  },
  TEARS: {
  	unicode: '\uD83D\uDE2D',
  	name: 'Tears'
  },
  INTERESTING: {
  	unicode: '\uD83E\uDD13',
  	name: 'Interesting'
  }
}

export const DAYS = ['M', 'T', 'W', 'R', 'F']

export const DRAGTYPES = {
  DRAG: 'drag', // drag a custom slot to a new location
  EXTEND: 'extend', // extend the length of a custom slot
  CREATE: 'create' // create a new custom slot
}
