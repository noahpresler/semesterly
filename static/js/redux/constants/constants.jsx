export const HALF_HOUR_HEIGHT = 25;
export const MAX_TIMETABLE_NAME_LENGTH = 30; // Length of "This is my fucking timetable!!"

export const DAYS = ['M', 'T', 'W', 'R', 'F'];

export const FULL_WEEK_LIST = ['U', 'M', 'T', 'W', 'R', 'F', 'S'];

export const VERBOSE_DAYS = ['Monday', 'Tuesday', 'Wednesday', 'Thursday', 'Friday'];

export const DRAG_TYPES = {
  DRAG: 'drag', // drag a custom slot to a new location
  EXTEND: 'extend', // extend the length of a custom slot
  CREATE: 'create', // create a new custom slot
};

// dictionary representing the order in which semesters occur
export const SEMESTER_RANKS = {
  Winter: 0,
  Spring: 1,
  Fall: 2,
};

export const CUSTOM_EVENT_FLAG = 'event';
