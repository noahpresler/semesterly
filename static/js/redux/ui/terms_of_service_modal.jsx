/*
Copyright (C) 2017 Semester.ly Technologies, LLC

Semester.ly is free software: you can redistribute it and/or modify
it under the terms of the GNU General Public License as published by
the Free Software Foundation, either version 3 of the License, or
(at your option) any later version.

Semester.ly is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU General Public License for more details.
*/

import React from 'react';
import PropTypes from 'prop-types';
import Modal from 'boron/WaveModal';

class TermsOfServiceModal extends React.Component {
  componentDidMount() {
    if (this.props.isVisible) {
      this.modal.show();
    }
  }

  componentDidUpdate() {
    if (this.props.isVisible) {
      this.modal.show();
    }
  }

  render() {
    const modalStyle = {
      width: '100%',
    };
    return (
      <Modal
        ref={(c) => { this.modal = c; }}
        className="terms-of-service-modal max-modal"
        modalStyle={modalStyle}
        closeOnClick={false}
      >
        <div className="tos-modal-container">
          <h1>Terms of Service and Privacy Policy</h1>
          <h3>
              Our Terms of Service and Privacy Policy have been updated. Please review them here:
          </h3>
          <div>
            <a
              href="/termsofservice"
              target="_blank"
              rel="noopener noreferrer"
              className="legal-links"
            >
              Terms of Service
              <i className="fa fa-external-link" />
            </a>
            <a
              href="/privacypolicy"
              target="_blank"
              rel="noopener noreferrer"
              className="legal-links"
            >
              Privacy Policy
              <i className="fa fa-external-link" />
            </a>
          </div>
          <button
            className="accept-tos-btn" onClick={() => {
              this.props.acceptTOS();
              this.modal.hide();
            }}
          >
            <i className="fa fa-check" />
            <span>I accept the Terms of Service</span>
          </button>
          <p className="method-details">
            You must accept the new Terms of Service to continue using Semester.ly.
          </p>
        </div>
      </Modal>
    );
  }
}

TermsOfServiceModal.propTypes = {
  isVisible: PropTypes.bool.isRequired,
  acceptTOS: PropTypes.func.isRequired,
};

export default TermsOfServiceModal;
