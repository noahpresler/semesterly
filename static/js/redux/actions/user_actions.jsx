/*
Copyright (C) 2017 Semester.ly Technologies, LLC

Semester.ly is free software: you can redistribute it and/or modify
it under the terms of the GNU General Public License as published by
the Free Software Foundation, either version 3 of the License, or
(at your option) any later version.

Semester.ly is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU General Public License for more details.
*/

import fetch from 'isomorphic-fetch';
import Cookie from 'js-cookie';
import uniq from 'lodash/uniq';
import {
    deleteRegistrationTokenEndpoint,
    getClassmatesEndpoint,
    getDeleteTimetableEndpoint,
    getFriendsEndpoint,
    getIntegrationEndpoint,
    getLoadSavedTimetablesEndpoint,
    getLogFacebookAlertClickEndpoint,
    getLogFacebookAlertViewEndpoint,
    getMostClassmatesCountEndpoint,
    getSaveSettingsEndpoint,
    getSaveTimetableEndpoint,
    getSetRegistrationTokenEndpoint,
    acceptTOSEndpoint,
} from '../constants/endpoints';
import {
  getActiveTimetable,
  getCurrentSemester } from '../reducers/root_reducer';
import { fetchCourseClassmates } from './modal_actions';
import { getNumberedName, loadTimetable, nullifyTimetable } from './timetable_actions';
import { receiveCourses } from './search_actions';
import { MAX_TIMETABLE_NAME_LENGTH } from '../constants/constants';
import * as ActionTypes from '../constants/actionTypes';
import { setTimeShownBanner, checkStatus, clearLocalTimetable } from '../util';

let autoSaveTimer;

export const receiveClassmates = json => dispatch => (
  dispatch({
    type: ActionTypes.CLASSMATES_RECEIVED,
    courses: json,
  })
);

export const getFriends = json => ({
  type: ActionTypes.FRIENDS_RECEIVED,
  peers: json,
});

export const requestClassmates = () => ({
  type: ActionTypes.REQUEST_CLASSMATES,
});

export const requestFriends = () => ({
  type: ActionTypes.REQUEST_FRIENDS,
});

const getSaveTimetablesRequestBody = (state) => {
  const tt = getActiveTimetable(state);
  // TODO: optional courses?
  return {
    slots: tt.slots,
    events: state.customSlots,
    has_conflict: tt.has_conflict,
    semester: getCurrentSemester(state),
    name: state.savingTimetable.activeTimetable.name,
    id: state.savingTimetable.activeTimetable.id,
  };
};

// returns the course to sections of a timetable
// TODO: when removing course to sections state, this should become the selector
export const lockActiveSections = (timetable) => {
  const courseSections = {};
  timetable.slots.forEach((slot) => {
    if (!(slot.course.id in courseSections)) {
      courseSections[slot.course.id] = {};
    }
    courseSections[slot.course.id][slot.section.section_type] = slot.section.meeting_section;
  });
  return courseSections;
};

export const requestMostClassmates = () => ({
  type: ActionTypes.REQUEST_MOST_CLASSMATES,
});

export const fetchMostClassmatesCount = timetable => (dispatch, getState) => {
  const state = getState();
  const courseIds = uniq(timetable.slots.map(s => s.course));

  if (!state.userInfo.data.social_courses) {
    return;
  }
  const semester = getCurrentSemester(state);
  dispatch(requestMostClassmates());
  fetch(getMostClassmatesCountEndpoint(semester, courseIds), {
    credentials: 'include',
    method: 'GET',
  })
      .then(response => response.json())
      .then((json) => {
        dispatch({
          type: ActionTypes.CHANGE_MOST_FRIENDS_CLASS,
          classId: json.id,
          count: json.count,
          total: json.total_count,
        });
      });
};

export const fetchClassmates = timetable => (dispatch, getState) => {
  const state = getState();
  const courseIds = uniq(timetable.slots.map(s => s.course));

  if (!state.userInfo.data.social_courses || courseIds.length === 0) {
    return;
  }
  setTimeout(() => {
    dispatch(fetchMostClassmatesCount(timetable));
  }, 500);
  dispatch(requestClassmates());
  fetch(getClassmatesEndpoint(getCurrentSemester(state), courseIds), {
    credentials: 'include',
    method: 'GET',
  })
    .then(response => response.json())
    .then((json) => {
      dispatch(receiveClassmates(json));
    });
};

export const saveTimetable = (isAutoSave = false, callback = null) => (dispatch, getState) => {
  const state = getState();
  if (!state.userInfo.data.isLoggedIn) {
    return dispatch({ type: ActionTypes.TOGGLE_SIGNUP_MODAL });
  }
  const activeTimetable = getActiveTimetable(state);

  // if current timetable is empty or we're already in saved state, don't save this timetable
  const numSlots = activeTimetable.slots.length + state.customSlots.length;
  if (numSlots === 0 || state.savingTimetable.upToDate) {
    return null;
  }

  // mark that we're now trying to save this timetable
  dispatch({
    type: ActionTypes.REQUEST_SAVE_TIMETABLE,
  });

  const body = getSaveTimetablesRequestBody(state);
  return fetch(getSaveTimetableEndpoint(), {
    headers: {
      'X-CSRFToken': Cookie.get('csrftoken'),
      Accept: 'application/json',
      'Content-Type': 'application/json',
    },
    method: 'POST',
    body: JSON.stringify(body),
    credentials: 'include',
  })
    .then(checkStatus)
    .then(response => response.json())
    .then((json) => {
      dispatch(loadTimetable(json.saved_timetable));
      dispatch({
        type: ActionTypes.RECEIVE_SAVED_TIMETABLES,
        timetables: json.timetables,
      });
      if (callback !== null) {
        callback();
      }
      return null;
    })
    .catch((error) => {
      if (error.response && error.response.status === 409) {
        dispatch({
          type: ActionTypes.ALERT_TIMETABLE_EXISTS,
        });
      }
      return null;
    });
};

export const duplicateTimetable = timetable => (dispatch, getState) => {
  const state = getState();
  if (!state.userInfo.data.isLoggedIn) {
    dispatch({ type: ActionTypes.TOGGLE_SIGNUP_MODAL });
  }
  // mark that we're now trying to save this timetable
  dispatch({
    type: ActionTypes.REQUEST_SAVE_TIMETABLE,
  });

  fetch(getSaveTimetableEndpoint(), {
    headers: {
      'X-CSRFToken': Cookie.get('csrftoken'),
      Accept: 'application/json',
      'Content-Type': 'application/json',
    },
    method: 'POST',
    body: JSON.stringify({
      semester: getCurrentSemester(state),
      source: timetable.name,
      name: getNumberedName(timetable.name, state.userInfo.data.timetables),
    }),
    credentials: 'include',
  })
  .then(response => response.json())
  .then((json) => {
    dispatch(loadTimetable(json.saved_timetable));
    dispatch({
      type: ActionTypes.RECEIVE_SAVED_TIMETABLES,
      timetables: json.timetables,
    });
    return json;
  });
};


export const deleteTimetable = timetable => (dispatch, getState) => {
  const state = getState();
  if (!state.userInfo.data.isLoggedIn) {
    dispatch({ type: ActionTypes.TOGGLE_SIGNUP_MODAL });
  }
    // mark that we're now trying to save this timetable
  dispatch({
    type: ActionTypes.REQUEST_SAVE_TIMETABLE,
  });
  fetch(getDeleteTimetableEndpoint(getCurrentSemester(state), timetable.name), {
    headers: {
      'X-CSRFToken': Cookie.get('csrftoken'),
      Accept: 'application/json',
      'Content-Type': 'application/json',
    },
    method: 'DELETE',
    credentials: 'include',
  })
    .then(response => response.json())
    .then((json) => {
      dispatch({
        type: ActionTypes.RECEIVE_SAVED_TIMETABLES,
        timetables: json.timetables,
      });
      if (json.timetables.length > 0) {
        dispatch(loadTimetable(json.timetables[0]));
      } else {
        nullifyTimetable(dispatch);
      }
      return json;
    });
};

export const saveSettings = callback => (dispatch, getState) => {
  dispatch({
    type: ActionTypes.REQUEST_SAVE_USER_INFO,
  });
  fetch(getSaveSettingsEndpoint(), {
    headers: {
      'X-CSRFToken': Cookie.get('csrftoken'),
      Accept: 'application/json',
      'Content-Type': 'application/json',
    },
    method: 'PATCH',
    body: JSON.stringify(getState().userInfo.data),
    credentials: 'include',
  })
    .then((response) => {
      const state = getState();
      if (state.userInfo.data.social_courses) {
        dispatch(fetchClassmates(getActiveTimetable(state)));
        if (state.courseInfo.id) {
          dispatch(fetchCourseClassmates(state.courseInfo.id));
        }
      }
      dispatch({
        type: ActionTypes.RECEIVE_USER_INFO_SAVED,
      });
      return response;
    })
    .then(() => {
      if (callback) {
        callback();
      }
    });
};

export const getUserSavedTimetables = semester => (dispatch) => {
  dispatch({
    type: ActionTypes.REQUEST_SAVE_USER_INFO,
  });
  fetch(getLoadSavedTimetablesEndpoint(semester), {
    credentials: 'include',
  })
    .then(response => response.json())
    .then((json) => {
      const { timetables, courses } = json;
      dispatch(receiveCourses(courses));
      dispatch({
        type: ActionTypes.RECEIVE_SAVED_TIMETABLES,
        timetables,
      });
      if (timetables[0]) {
        dispatch(loadTimetable(timetables[0]));
      } else {
        dispatch(nullifyTimetable(dispatch));
      }
    });
};

export const fetchFriends = () => (dispatch, getState) => {
  const state = getState();
  if (!state.userInfo.data.social_courses) {
    return;
  }
  dispatch(requestFriends());
  dispatch({
    type: ActionTypes.PEER_MODAL_LOADING,
  });
  fetch(getFriendsEndpoint(getCurrentSemester(state)), {
    credentials: 'include',
    method: 'GET',
  })
    .then(response => response.json())
    .then((json) => {
      dispatch(getFriends(json));
      dispatch({
        type: ActionTypes.PEER_MODAL_LOADED,
      });
    });
};

export const autoSave = (delay = 2000) => (dispatch, getState) => {
  clearTimeout(autoSaveTimer);
  autoSaveTimer = setTimeout(() => {
    const state = getState();
    const existsSlots = getActiveTimetable(state).slots.length > 0;
    const existsCustomEvents = state.customSlots.length > 0;
    if (state.userInfo.data.isLoggedIn && (existsSlots || existsCustomEvents)) {
      dispatch(saveTimetable(true));
      clearLocalTimetable();
    }
  }, delay);
};

export const sendRegistrationToken = token => (dispatch) => {
  fetch(getSetRegistrationTokenEndpoint(), {
    headers: {
      'X-CSRFToken': Cookie.get('csrftoken'),
      Accept: 'application/json',
      'Content-Type': 'application/json',
    },
    method: 'POST',
    body: JSON.stringify({
      token,
    }),
    credentials: 'include',
  })
    .then((response) => {
      if (response.status === 201) {
        dispatch({
          type: ActionTypes.TOKEN_REGISTERED,
        });
      }
    });
};

export const setARegistrationToken = () => (dispatch) => {
  if ('serviceWorker' in navigator) {
    navigator.serviceWorker.register('/sw.js').then((reg) => {
      reg.pushManager.subscribe({
        userVisibleOnly: true,
      }).then((sub) => {
        dispatch(sendRegistrationToken(sub.toJSON()));
      });
    }).catch(() => {
    });
  }
};

export const isRegistered = () => (dispatch) => {
  if ('serviceWorker' in navigator) {
    navigator.serviceWorker.register('/sw.js')
      .then(reg => reg.pushManager.getSubscription().then((sub) => {
        if (sub) {
          dispatch({
            type: ActionTypes.TOKEN_REGISTERED,
          });
          return true;
        }
        return null;
      })).catch(() => null);
  }
};

export const sendRegistrationTokenForDeletion = token => (dispatch) => {
  fetch(deleteRegistrationTokenEndpoint(token.endpoint), {
    headers: {
      'X-CSRFToken': Cookie.get('csrftoken'),
      Accept: 'application/json',
      'Content-Type': 'application/json',
    },
    method: 'DELETE',
    credentials: 'include',
  })
    .then((response) => {
      if (response.status === 204) {
        dispatch({
          type: ActionTypes.UNREGISTER_TOKEN,
        });
      }
    });
};

export const unRegisterAToken = () => (dispatch) => {
  if ('serviceWorker' in navigator) {
    navigator.serviceWorker.register('/sw.js').then((reg) => {
      reg.pushManager.subscribe({
        userVisibleOnly: true,
      }).then((sub) => {
        dispatch(sendRegistrationTokenForDeletion(sub.toJSON()));
      });
    }).catch(() => {
    });
  }
};

export const openIntegrationModal = (integrationID, courseID) => (dispatch) => {
  fetch(getIntegrationEndpoint(integrationID, courseID), {
    credentials: 'include',
    method: 'GET',
  })
    .then((response) => {
      dispatch({
        type: ActionTypes.OPEN_INTEGRATION_MODAL,
        enabled: response.status === 200,
        id: courseID,
        integration_id: integrationID,
      });
    });
};

export const deleteUser = () => (dispatch) => {
  fetch(getSaveSettingsEndpoint(), {
    headers: {
      'X-CSRFToken': Cookie.get('csrftoken'),
      Accept: 'application/json',
      'Content-Type': 'application/json',
    },
    method: 'DELETE',
    credentials: 'include',
  })
    .then((response) => {
      dispatch({
        type: ActionTypes.DELETED_ACCOUNT,
        status: response.status,
      });
    });
};

export const delIntegration = (integrationID, courseID) => {
  fetch(getIntegrationEndpoint(integrationID, courseID), {
    headers: {
      'X-CSRFToken': Cookie.get('csrftoken'),
      Accept: 'application/json',
      'Content-Type': 'application/json',
    },
    credentials: 'include',
    method: 'DELETE',
  });
};

export const addIntegration = (integrationID, courseID, json) => {
  fetch(getIntegrationEndpoint(integrationID, courseID), {
    headers: {
      'X-CSRFToken': Cookie.get('csrftoken'),
      Accept: 'application/json',
      'Content-Type': 'application/json',
    },
    credentials: 'include',
    method: 'POST',
    body: JSON.stringify({ json }),
  });
};

export const logFacebookAlertView = () => {
  fetch(getLogFacebookAlertViewEndpoint(), {
    method: 'POST',
    credentials: 'include',
  });
};

export const LogFacebookAlertClick = () => {
  fetch(getLogFacebookAlertClickEndpoint(), {
    method: 'POST',
    credentials: 'include',
  });
};

export const changeUserInfo = info => ({
  type: ActionTypes.CHANGE_USER_INFO,
  data: info,
});

export const changeTimetableName = name => (dispatch) => {
  if (name.length === 0 || name.length > MAX_TIMETABLE_NAME_LENGTH) {
    return;
  }
  dispatch({
    type: ActionTypes.CHANGE_ACTIVE_SAVED_TIMETABLE_NAME,
    name,
  });
  dispatch(saveTimetable());
};

export const acceptTOS = () => (dispatch) => {
  fetch(acceptTOSEndpoint(), {
    headers: {
      'X-CSRFToken': Cookie.get('csrftoken'),
      Accept: 'application/json',
      'Content-Type': 'application/json',
    },
    credentials: 'include',
    method: 'POST',
    body: '',
  })
    .then((response) => {
      if (response.status === 204) {
        dispatch({
          type: ActionTypes.CLOSE_TOS_MODAL,
        });
      }
    });
};

// Show the TOS and privacy policy agreement if the user has not seen the latest version.
// The modal is used for logged in users and the banner is used for anonymous users.
export const handleAgreement = (currentUser, timeUpdatedTos) => (dispatch) => {
  if (currentUser.isLoggedIn) {
    const timeAcceptedTos = currentUser.timeAcceptedTos;
    if (!timeAcceptedTos || Date.parse(timeAcceptedTos) < timeUpdatedTos) {
      dispatch({ type: ActionTypes.TRIGGER_TOS_MODAL });
    }
  } else {
    const timeShownBanner = localStorage.getItem('timeShownBanner');
    if (!timeShownBanner || timeShownBanner < timeUpdatedTos) {
      setTimeShownBanner(Date.now());
      dispatch({ type: ActionTypes.TRIGGER_TOS_BANNER });
    }
  }
};
