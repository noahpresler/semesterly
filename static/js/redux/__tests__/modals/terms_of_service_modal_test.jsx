/*
Copyright (C) 2017 Semester.ly Technologies, LLC

Semester.ly is free software: you can redistribute it and/or modify
it under the terms of the GNU General Public License as published by
the Free Software Foundation, either version 3 of the License, or
(at your option) any later version.

Semester.ly is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU General Public License for more details.
*/

import React from 'react';
import thunk from 'redux-thunk';
import Provider from 'react-redux/src/components/Provider';
import configureMockStore from 'redux-mock-store';
import renderer from 'react-test-renderer';
import { tosModalFixture, userInfoFixture } from '../../__fixtures__/terms_of_service_modal.fixture';
import TermsOfServiceModalContainer from '../../ui/containers/terms_of_service_modal_container';
import { handleAgreement } from '../../actions/user_actions';
import * as ActionTypes from '../../constants/actionTypes';

const middlewares = [thunk];
const mockStore = configureMockStore(middlewares);

describe('TOS Modal', () => {
  it('shows when isVisible is true', () => {
    const store = mockStore({
      termsOfServiceModal: tosModalFixture,
      userInfo: userInfoFixture,
    });
    const tree = renderer.create(
      <Provider store={store}><TermsOfServiceModalContainer /></Provider>,
    ).toJSON();
    delete tree.children[0].children[0].props.style.animationName;
    delete tree.children[1].props.style.animationName;
    expect(tree).toMatchSnapshot();
  });

  it('shows welcome message for new users', () => {
    const newUser = userInfoFixture;
    newUser.data.timeAcceptedTos = null;
    const store = mockStore({
      termsOfServiceModal: tosModalFixture,
      userInfo: userInfoFixture,
    });
    const tree = renderer.create(
      <Provider store={store}><TermsOfServiceModalContainer /></Provider>,
    ).toJSON();
    delete tree.children[0].children[0].props.style.animationName;
    delete tree.children[1].props.style.animationName;
    expect(tree).toMatchSnapshot();
  });

  it('is hidden when isVisible is false', () => {
    const store = mockStore({
      termsOfServiceModal: { ...tosModalFixture, isVisible: false },
      userInfo: userInfoFixture,
    });
    const tree = renderer.create(
      <Provider store={store}><TermsOfServiceModalContainer /></Provider>,
    ).toJSON();
    expect(tree).toMatchSnapshot();
  });
});

describe('HandleAgreement correctly triggers tosMODAL', () => {
  it('if unaccepted', () => {
    const store = mockStore({
      termsOfServiceModal: tosModalFixture,
      userInfo: userInfoFixture,
    });

    const currentUser = {
      timeAcceptedTos: null,
      isLoggedIn: true,
    };

    store.dispatch(handleAgreement(currentUser, Date.now()));
    const expectedActions = store.getActions();
    expect(expectedActions[0]).toEqual({ type: ActionTypes.TRIGGER_TOS_MODAL });
  });

  it('if accepted but outdated', () => {
    const store = mockStore({
      termsOfServiceModal: tosModalFixture,
      userInfo: userInfoFixture,
    });

    const currentUser = {
      timeAcceptedTos: 0,
      isLoggedIn: true,
    };

    store.dispatch(handleAgreement(currentUser, Date.parse(1)));
    const expectedActions = store.getActions();
    expect(expectedActions[0]).toEqual({ type: ActionTypes.TRIGGER_TOS_MODAL });
  });

  it('empty if logged in and accepted', () => {
    const store = mockStore({
      termsOfServiceModal: tosModalFixture,
      userInfo: userInfoFixture,
    });

    const currentUser = {
      timeAcceptedTos: 2,
      isLoggedIn: true,
    };

    store.dispatch(handleAgreement(currentUser, Date.parse(1)));
    const expectedActions = store.getActions();
    expect(expectedActions).toEqual([]);
  });
});
