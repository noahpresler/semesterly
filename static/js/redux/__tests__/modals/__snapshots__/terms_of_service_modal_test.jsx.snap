// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TOS Modal is hidden when isVisible is false 1`] = `null`;

exports[`TOS Modal shows welcome message for new users 1`] = `
<span>
  <div
    className="terms-of-service-modal max-modal"
    style={
      Object {
        "left": "50%",
        "position": "fixed",
        "top": "50%",
        "transform": "translate3d(-50%, -50%, 0)",
        "width": "100%",
        "zIndex": 1050,
      }
    }
  >
    <div
      style={
        Object {
          "animationDuration": "1s",
          "animationFillMode": "forwards",
          "animationTimingFunction": "linear",
          "backgroundColor": "white",
          "margin": 0,
        }
      }
      tabIndex="-1"
    >
      <div
        className="tos-modal-container"
      >
        <h1>
          Terms of Service and Privacy Policy
        </h1>
        <h3>
          Welcome to Semester.ly! Please take a look at our Terms of Service and Privacy Policy before getting started:
        </h3>
        <div>
          <a
            className="legal-links"
            href="/termsofservice"
            rel="noopener noreferrer"
            target="_blank"
          >
            Terms of Service
            <i
              className="fa fa-external-link"
            />
          </a>
          <a
            className="legal-links"
            href="/privacypolicy"
            rel="noopener noreferrer"
            target="_blank"
          >
            Privacy Policy
            <i
              className="fa fa-external-link"
            />
          </a>
        </div>
        <button
          className="accept-tos-btn"
          onClick={[Function]}
        >
          <i
            className="fa fa-check"
          />
          <span>
            I accept the Terms of Service
          </span>
        </button>
        <p
          className="method-details"
        >
          You must accept the new Terms of Service to continue using Semester.ly.
        </p>
      </div>
    </div>
  </div>
  <div
    onClick={null}
    style={
      Object {
        "animationDuration": "0.3s",
        "animationFillMode": "forwards",
        "animationTimingFunction": "linear",
        "backgroundColor": "#373A47",
        "bottom": 0,
        "left": 0,
        "position": "fixed",
        "right": 0,
        "top": 0,
        "zIndex": 1040,
      }
    }
  />
</span>
`;

exports[`TOS Modal shows when isVisible is true 1`] = `
<span>
  <div
    className="terms-of-service-modal max-modal"
    style={
      Object {
        "left": "50%",
        "position": "fixed",
        "top": "50%",
        "transform": "translate3d(-50%, -50%, 0)",
        "width": "100%",
        "zIndex": 1050,
      }
    }
  >
    <div
      style={
        Object {
          "animationDuration": "1s",
          "animationFillMode": "forwards",
          "animationTimingFunction": "linear",
          "backgroundColor": "white",
          "margin": 0,
        }
      }
      tabIndex="-1"
    >
      <div
        className="tos-modal-container"
      >
        <h1>
          Terms of Service and Privacy Policy
        </h1>
        <h3>
          Our Terms of Service and Privacy Policy have been updated. Please review them here:
        </h3>
        <div>
          <a
            className="legal-links"
            href="/termsofservice"
            rel="noopener noreferrer"
            target="_blank"
          >
            Terms of Service
            <i
              className="fa fa-external-link"
            />
          </a>
          <a
            className="legal-links"
            href="/privacypolicy"
            rel="noopener noreferrer"
            target="_blank"
          >
            Privacy Policy
            <i
              className="fa fa-external-link"
            />
          </a>
        </div>
        <button
          className="accept-tos-btn"
          onClick={[Function]}
        >
          <i
            className="fa fa-check"
          />
          <span>
            I accept the Terms of Service
          </span>
        </button>
        <p
          className="method-details"
        >
          You must accept the new Terms of Service to continue using Semester.ly.
        </p>
      </div>
    </div>
  </div>
  <div
    onClick={null}
    style={
      Object {
        "animationDuration": "0.3s",
        "animationFillMode": "forwards",
        "animationTimingFunction": "linear",
        "backgroundColor": "#373A47",
        "bottom": 0,
        "left": 0,
        "position": "fixed",
        "right": 0,
        "top": 0,
        "zIndex": 1040,
      }
    }
  />
</span>
`;
